Mathlib/Analysis/NormedSpace/HahnBanach/Separation.lean

```diff
@@ -200,3 +202,103 @@ theorem iInter_halfspaces_eq (hs₁ : Convex ℝ s) (hs₂ : IsClosed s) :
+
+/--Real linear extension of continuous extension of `LinearMap.extendTo𝕜'` -/
+noncomputable def LinTo𝕜' : (E →L[ℝ] ℝ) →ₗ[ℝ] (E →L[𝕜] 𝕜) :=
+  letI to𝕜 (fr : (E →L[ℝ] ℝ)) : (E →L[𝕜] 𝕜) :=
+    { toLinearMap := LinearMap.extendTo𝕜' fr
+      cont := show Continuous fun x ↦ (fr x : 𝕜) - (I : 𝕜) * (fr ((I : 𝕜) • x) : 𝕜) by fun_prop }
+  have h fr x : to𝕜 fr x = ((fr x : 𝕜) - (I : 𝕜) * (fr ((I : 𝕜) • x) : 𝕜)) := rfl
+  { toFun := to𝕜
+    map_add' := by intros; ext; simp [h]; ring
+    map_smul' := by intros; ext; simp [h, real_smul_eq_coe_mul]; ring }
+
+@[simp]
+lemma real_of_LinTo𝕜' (g : E →L[ℝ] ℝ) (x : E) :  re ((LinTo𝕜' g) x : 𝕜) = g x := by
+  have h g (x : E) : LinTo𝕜' g x = ((g x : 𝕜) - (I : 𝕜) * (g ((I : 𝕜) • x) : 𝕜)) := rfl
+  simp only [h , map_sub, ofReal_re, mul_re, I_re, zero_mul, ofReal_im, mul_zero,
+  sub_self, sub_zero]
+
+variable [ContinuousSMul ℝ E]
+
+theorem separate_convex_open_set_RCLike  {s : Set E}
+    (hs₀ : (0 : E) ∈ s) (hs₁ : Convex ℝ s) (hs₂ : IsOpen s) {x₀ : E} (hx₀ : x₀ ∉ s) :
+    ∃ f : E →L[𝕜] 𝕜, re (f x₀) = 1 ∧ ∀ x ∈ s, re (f x) < 1 := by
+  obtain ⟨g, hg⟩ := separate_convex_open_set hs₀ hs₁ hs₂ hx₀; use LinTo𝕜' g
+  simp only [real_of_LinTo𝕜']; exact hg
+
+theorem geometric_hahn_banach_open_RCLike (hs₁ : Convex ℝ s) (hs₂ : IsOpen s) (ht : Convex ℝ t)
+    (disj : Disjoint s t) : ∃ (f : E →L[𝕜] 𝕜) (u : ℝ), (∀ a ∈ s, re (f a) < u) ∧
+    ∀ b ∈ t, u ≤ re (f b) := by
+  obtain ⟨f, u, h⟩ := geometric_hahn_banach_open hs₁ hs₂ ht disj; use LinTo𝕜' f
+  simp only [real_of_LinTo𝕜']; exact Exists.intro u h
+
+theorem geometric_hahn_banach_open_point_RCLike (hs₁ : Convex ℝ s) (hs₂ : IsOpen s) (disj : x ∉ s) :
+    ∃ f : E →L[𝕜] 𝕜, ∀ a ∈ s, re (f a) < re (f x) := by
+  obtain ⟨f, h⟩ := geometric_hahn_banach_open_point hs₁ hs₂ disj; use LinTo𝕜' f
+  simp only [real_of_LinTo𝕜']; exact fun a a_1 ↦ h a a_1
+
+theorem geometric_hahn_banach_point_open_RCLike (ht₁ : Convex ℝ t) (ht₂ : IsOpen t) (disj : x ∉ t) :
+    ∃ f : E →L[𝕜] 𝕜, ∀ b ∈ t, re (f x) < re (f b) :=
+  let ⟨f, hf⟩ := geometric_hahn_banach_open_point_RCLike ht₁ ht₂ disj
+  ⟨-f, by simpa⟩
+
+theorem geometric_hahn_banach_open_open_RCLike (hs₁ : Convex ℝ s) (hs₂ : IsOpen s)
+    (ht₁ : Convex ℝ t) (ht₃ : IsOpen t) (disj : Disjoint s t) :
+    ∃ (f : E →L[𝕜] 𝕜) (u : ℝ), (∀ a ∈ s, re (f a) < u) ∧ ∀ b ∈ t, u < re (f b) := by
+  obtain ⟨f, u, h⟩ := geometric_hahn_banach_open_open hs₁ hs₂ ht₁ ht₃ disj
+  use LinTo𝕜' f; simp only [real_of_LinTo𝕜']; exact Exists.intro u h
+
+variable [LocallyConvexSpace ℝ E]
+
+theorem geometric_hahn_banach_compact_closed_RCLike (hs₁ : Convex ℝ s) (hs₂ : IsCompact s)
+    (ht₁ : Convex ℝ t) (ht₂ : IsClosed t) (disj : Disjoint s t) :
+    ∃ (f : E →L[𝕜] 𝕜) (u v : ℝ), (∀ a ∈ s, re (f a) < u) ∧ u < v ∧ ∀ b ∈ t, v < re (f b) := by
+  obtain ⟨g, u, v, h1⟩ := geometric_hahn_banach_compact_closed hs₁ hs₂ ht₁ ht₂ disj; use LinTo𝕜' g
+  simp only [real_of_LinTo𝕜', exists_and_left]; use u; constructor; exact h1.1; use v
+  exact h1.2
```

Comment:
Probably this works:
```suggestion
  simp only [real_of_LinTo𝕜', exists_and_left]
  exact ⟨u, h1⟩
```
or this:
```suggestion
  simp only [real_of_LinTo𝕜', exists_and_left]
  exact ⟨u, h1.1, h1.2⟩
```

Reply:
Somewhat irritatingly, neither works.

Reply:
Thanks, by the way, for looking over this!!

Reply:
`exact ⟨u, h1.1, v, h1.2⟩` does the job!